File: #2 Singly Linkedlist/singly_linkedlist.py
class Node:
    def __init__(self, nim, nama):
        self.nim = nim
        self.nama = nama
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.count = 0

    def insert_beginning(self, nim, nama):
        new = Node(nim, nama)
        new.next = self.head
        self.head = new
        self.count += 1

    def insert_end(self, nim, nama):
        new = Node(nim, nama)
        if not self.head:
            self.head = new
        else:
            temp = self.head
            while temp.next:
                temp = temp.next
            temp.next = new
        self.count += 1

    def insert_position(self, nim, nama, pos):
        if pos < 1 or pos > self.count + 1:
            print("Posisi tidak valid")
            return
        if pos == 1:
            self.insert_beginning(nim, nama)
            return
        new = Node(nim, nama)
        temp = self.head
        for _ in range(pos - 2):
            temp = temp.next
        new.next = temp.next
        temp.next = new
        self.count += 1

    def delete_beginning(self):
        if self.head:
            self.head = self.head.next
            self.count -= 1

    def delete_end(self):
        if not self.head:
            return
        if not self.head.next:
            self.head = None
        else:
            temp = self.head
            while temp.next.next:
                temp = temp.next
            temp.next = None
        self.count -= 1

    def delete_position(self, pos):
        if pos < 1 or pos > self.count:
            print("Posisi tidak valid")
            return
        if pos == 1:
            self.delete_beginning()
            return
        temp = self.head
        for _ in range(pos - 2):
            temp = temp.next
        temp.next = temp.next.next
        self.count -= 1

    def delete_by_nim(self, nim):
        if not self.head:
            return
        if self.head.nim == nim:
            self.head = self.head.next
            self.count -= 1
            return
        temp = self.head
        while temp.next and temp.next.nim != nim:
            temp = temp.next
        if temp.next:
            temp.next = temp.next.next
            self.count -= 1

    def show(self):
        temp = self.head
        i = 1
        while temp:
            print(f"{i}. {temp.nim} - {temp.nama}")
            temp = temp.next
            i += 1

ll = LinkedList()

while True:
    print("\nMENU")
    print("1. Insert Beginning")
    print("2. Insert Position")
    print("3. Insert End")
    print("4. Delete Beginning")
    print("5. Delete Position")
    print("6. Delete End")
    print("7. Delete by NIM")
    print("8. Show Data")
    print("9. Exit")

    ch = int(input("Pilih: "))

    if ch == 1:
        nim = input("NIM: ")
        nama = input("Nama: ")
        ll.insert_beginning(nim, nama)
    elif ch == 2:
        pos = int(input("Posisi: "))
        nim = input("NIM: ")
        nama = input("Nama: ")
        ll.insert_position(nim, nama, pos)
    elif ch == 3:
        nim = input("NIM: ")
        nama = input("Nama: ")
        ll.insert_end(nim, nama)
    elif ch == 4:
        ll.delete_beginning()
    elif ch == 5:
        pos = int(input("Posisi: "))
        ll.delete_position(pos)
    elif ch == 6:
        ll.delete_end()
    elif ch == 7:
        nim = input("NIM: ")
        ll.delete_by_nim(nim)
    elif ch == 8:
        ll.show()
    elif ch == 9:
        break

File: #2 Singly Linkedlist/singly_linked_list.java
import java.util.Scanner;

class Node {
    String nim, nama;
    Node next;

    Node(String nim, String nama) {
        this.nim = nim;
        this.nama = nama;
        this.next = null;
    }
}

public class singly_linked_list {
    static Node head = null;
    static int count = 0;

    static void insertBeginning(String nim, String nama) {
        Node newNode = new Node(nim, nama);
        newNode.next = head;
        head = newNode;
        count++;
    }

    static void insertEnd(String nim, String nama) {
        Node newNode = new Node(nim, nama);
        if (head == null) {
            head = newNode;
        } else {
            Node temp = head;
            while (temp.next != null) temp = temp.next;
            temp.next = newNode;
        }
        count++;
    }

    static void insertPosition(String nim, String nama, int pos) {
        if (pos < 1 || pos > count + 1) {
            System.out.println("Posisi tidak valid");
            return;
        }
        if (pos == 1) {
            insertBeginning(nim, nama);
            return;
        }
        Node newNode = new Node(nim, nama);
        Node temp = head;
        for (int i = 1; i < pos - 1; i++) temp = temp.next;
        newNode.next = temp.next;
        temp.next = newNode;
        count++;
    }

    static void deleteBeginning() {
        if (head == null) return;
        head = head.next;
        count--;
    }

    static void deleteEnd() {
        if (head == null) return;
        if (head.next == null) {
            head = null;
        } else {
            Node temp = head;
            while (temp.next.next != null) temp = temp.next;
            temp.next = null;
        }
        count--;
    }

    static void deletePosition(int pos) {
        if (pos < 1 || pos > count) {
            System.out.println("Posisi tidak valid");
            return;
        }
        if (pos == 1) {
            deleteBeginning();
            return;
        }
        Node temp = head;
        for (int i = 1; i < pos - 1; i++) temp = temp.next;
        temp.next = temp.next.next;
        count--;
    }

    static void deleteByNim(String nim) {
        if (head == null) return;
        if (head.nim.equals(nim)) {
            head = head.next;
            count--;
            return;
        }
        Node temp = head;
        while (temp.next != null && !temp.next.nim.equals(nim)) temp = temp.next;
        if (temp.next != null) {
            temp.next = temp.next.next;
            count--;
        }
    }

    static void showData() {
        Node temp = head;
        int i = 1;
        while (temp != null) {
            System.out.println(i + ". " + temp.nim + " - " + temp.nama);
            temp = temp.next;
            i++;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int choice;

        do {
            System.out.println("\nMENU:");
            System.out.println("1. Insert Beginning");
            System.out.println("2. Insert Position");
            System.out.println("3. Insert End");
            System.out.println("4. Delete Beginning");
            System.out.println("5. Delete Position");
            System.out.println("6. Delete End");
            System.out.println("7. Delete by NIM");
            System.out.println("8. Show Data");
            System.out.println("9. Exit");
            System.out.print("Pilih: ");
            choice = sc.nextInt();
            sc.nextLine();

            String nim, nama;
            int pos;

            switch (choice) {
                case 1:
                    System.out.print("NIM: "); nim = sc.nextLine();
                    System.out.print("Nama: "); nama = sc.nextLine();
                    insertBeginning(nim, nama);
                    break;
                case 2:
                    System.out.print("Posisi: "); pos = sc.nextInt(); sc.nextLine();
                    System.out.print("NIM: "); nim = sc.nextLine();
                    System.out.print("Nama: "); nama = sc.nextLine();
                    insertPosition(nim, nama, pos);
                    break;
                case 3:
                    System.out.print("NIM: "); nim = sc.nextLine();
                    System.out.print("Nama: "); nama = sc.nextLine();
                    insertEnd(nim, nama);
                    break;
                case 4: deleteBeginning(); break;
                case 5:
                    System.out.print("Posisi: "); pos = sc.nextInt();
                    deletePosition(pos);
                    break;
                case 6: deleteEnd(); break;
                case 7:
                    System.out.print("NIM: "); nim = sc.nextLine();
                    deleteByNim(nim);
                    break;
                case 8: showData(); break;
            }
        } while (choice != 9);
    }
}